forge-gui/src/main/java/forge/player/HumanPlaySpellAbility.java

```diff
@@ -162,19 +162,33 @@ public final boolean playAbility(final boolean mayChooseTargets, final boolean i
         // This line makes use of short-circuit evaluation of boolean values, that is each subsequent argument
         // is only executed or evaluated if the first argument does not suffice to determine the value of the expression
         // because of Selective Snare do announceType first
-        final boolean prerequisitesMet = announceType()
-                && announceValuesLikeX()
-                && ability.checkRestrictions(human)
-                && (!mayChooseTargets || ability.setupTargets()) // if you can choose targets, then do choose them.
-                && ability.canCastTiming(human)
-                && ability.isLegalAfterStack()
-                && (isFree || payment.payCost(new HumanCostDecision(controller, human, ability, ability.isTrigger())));
+
+        // We need to split up this super conditional to know WHY somethings prequisites failed.
+        // Failing because there's no legal targets is different than failing because the player canceled paying costs
+
+        final boolean announcePrequiste = announceType() && announceValuesLikeX();
```

Comment:
that's more expensive now and also will still let you choose targets even if you already canceled by announcing?
this should cascade in a cleaner way

Reply:
Fair enough. I think i can flatten most of these again. At the time i was trying to figure out where the rollbacks were failing.

Reply:
not handled yet:
https://stackoverflow.com/questions/3152208/do-and-short-circuit-in-java

Reply:
I hate Java.